dec/decode.go

```diff
@@ -65,3 +67,102 @@ func DecompressBuffer(encodedBuffer []byte, decodedBuffer []byte) ([]byte, error
 func toC(array []byte) *C.uint8_t {
 	return (*C.uint8_t)(unsafe.Pointer(&array[0]))
 }
+
+// Decompresses a Brotli-encoded stream using the io.Reader interface
+type BrotliReader struct {
+	state  C.BrotliState
+	reader io.Reader
+
+	// Internal buffer for compressed data
+	buffer []byte
+
+	availableIn C.size_t
+	nextIn      *C.uint8_t
+	totalOut    C.size_t
+}
+
+// Fill a buffer, p, with the decompressed contents of the stream.
+// Returns the number of bytes read, or an error
+func (r *BrotliReader) Read(p []byte) (int, error) {
+	var err error
+
+	// Prepare arguments
+	maxOutput := len(p)
+	availableOut := C.size_t(maxOutput)
+	nextOut := (*C.uint8_t)(unsafe.Pointer(&p[0]))
+	var read int
+
+	for availableOut > 0 {
+
+		// Read more compressed data
+		if r.availableIn == 0 {
+			var read int
+			if read, err = r.reader.Read(r.buffer); err != nil {
+				return 0, err
+			}
+			r.availableIn = C.size_t(read)
+			r.nextIn = (*C.uint8_t)(unsafe.Pointer(&r.buffer[0]))
+		}
+
+		if r.availableIn > 0 {
+			// Decompress
+			result := C.BrotliDecompressStream(
+				&r.availableIn,
+				&r.nextIn,
+				&availableOut,
+				&nextOut,
+				&r.totalOut,
+				&r.state,
+			)
+
+			read = maxOutput - int(availableOut)
+
+			switch result {
+			case C.BROTLI_RESULT_SUCCESS:
+				break
+			case C.BROTLI_RESULT_NEEDS_MORE_OUTPUT:
+				if read > 0 {
+					return read, nil
+				} else {
+					return 0, errors.New("Brotli decompression error: needs more output buffer")
+				}
+			case C.BROTLI_RESULT_ERROR:
+				return 0, errors.New("Brotli decompression error")
+			case C.BROTLI_RESULT_NEEDS_MORE_INPUT:
+				continue
+			default:
+				return 0, errors.New("Unrecognised Brotli decompression error")
+			}
+		}
+	}
+
+	return read, nil
+}
+
+func (r *BrotliReader) Close() error {
+
+	C.BrotliStateCleanup(&r.state)
+
+	if v, ok := r.reader.(io.Closer); ok {
+		return v.Close()
+	}
+
+	return nil
+}
+
+// Returns a Reader that decompresses the stream from another reader.
+// Ensure that you Close the stream when you are finished in order to clean up the
+// Brotli compression state
+func Decompress(stream io.Reader) io.ReadCloser {
+
+	r := &BrotliReader{
```

Comment:
> but I feel that the Go community isn't one to care about binary size that much.

*sigh* unfortunately, yep

> This means that we would have to expose the access through some wrapper in the enc and dec packages (maybe a type declaration is enough?) which would then allow the structure to be used from the main brotli package.

yep, that was exactly my line of thinking, it's going to get complicated for nothing, come to think of it I'm for keeping it simple with enc/dec separate - if users have conflicting packages they can always alias them as import time.

> Regarding API breakage  / Suggested plan

sounds good, although I'd err on the side of keeping brotli-go v0 for some more time - as far as I can tell, `v0` doesn't in fact guarantee a stable interface, just that stuff is brewing - `v1` is the first stable (at least that's how it works in semver).

I feel like it'd be good to have:

  * stream support
  * travis-ci
  * improved README

...all still as v0 for a little while (in fact, none of those break the API currently available on master?), and then see if we run into issues using brotli-go in other projects (I'm planning on using it both as transport compression in a rsync-like protocol, and as offline compression for distributing delta updates, so hopefully that's real-world enough usage to uncover obvious issues)